src/DynamoCore/Utilities/NodeModelExtensions.cs

```diff
@@ -0,0 +1,132 @@
+                gathered.Add(n);
+            }
+
+            foreach (var n in upstream)
+            {
+                n.UpstreamNodesMatchingPredicate(gathered, match);
+            }
+
+            return gathered;
+        }
+
+        internal static IEnumerable<NodeModel> AllUpstreamNodes(this NodeModel node, List<NodeModel> gathered)
+        {
+            var upstream = node.InPorts.SelectMany(p => p.Connectors.Select(c => c.Start.Owner));
+
+            foreach (var n in upstream.Where(n => !gathered.Contains(n)))
+            {
+                gathered.Add(n);
+            }
+
+            foreach (var n in upstream)
+            {
+                n.AllUpstreamNodes(gathered);
+            }
+
+            return gathered;
+        }
+
+        internal static bool IsUpstreamOf(this NodeModel node, NodeModel otherNode)
+        {
+            var gathered = new List<NodeModel>();
+            otherNode.AllUpstreamNodes(gathered);
+
+            return gathered.Contains(node);
+        }
+
+        public static List<IGraphicItem> GeneratedGraphicItems(this NodeModel node, EngineController engineController)
+        {
+            var ids = node.GetAllOutportAstIdentifiers();
+
+            var results = new List<IGraphicItem>();
+
+            foreach (var id in ids)
+            {
+                var mirror = engineController.GetMirror(id);
+                if (mirror == null) continue;
+
+                var mirrorData = mirror.GetData();
+                if (mirrorData == null) continue;
+
+                GetGraphicItemsFromMirrorData(mirrorData, results);
+            }
+
+            return results;
+        }
+
+        private static void GetGraphicItemsFromMirrorData(MirrorData mirrorData, List<IGraphicItem> graphicItems)
+        {
+            if (mirrorData == null) return;
+
+            if (mirrorData.IsCollection)
+            {
+                foreach (var el in mirrorData.GetElements())
+                {
+                    GetGraphicItemsFromMirrorData(el, graphicItems);
+                }
+            }
+            else
+            {
+                var graphicItem = mirrorData.Data as IGraphicItem;
+                if (graphicItem == null) return;
+
+                graphicItems.Add(graphicItem);
+            }
+        }
+
+        private static IEnumerable<string> GetAllOutportAstIdentifiers(this NodeModel node)
+        {
+            var ids = new List<string>();
+            for (var i = 0; i < node.OutPortData.Count; ++i)
+            {
+                var id = node.GetOneOutportAstIdentifier(i);
+                if (!string.IsNullOrEmpty(id))
+                    ids.Add(id);
+            }
+
+            return ids;
+        }
+
+        private static string GetOneOutportAstIdentifier(this NodeModel node, int outPortIndex)
+        {
+            var output = node.GetAstIdentifierForOutputIndex(outPortIndex);
+            return output == null ? null : output.ToString();
+        }
+    }
```

Comment:
I'm guessing these are moved from elsewhere? I'm guessing you have some tests for these?